osu.Game/Screens/OnlinePlay/Multiplayer/Match/Playlist/MultiplayerQueueList.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Rooms;
+using osuTK;
+
+namespace osu.Game.Screens.OnlinePlay.Multiplayer.Match.Playlist
+{
+    /// <summary>
+    /// A gameplay-ordered list of <see cref="DrawableRoomPlaylistItem"/>s.
+    /// </summary>
+    public class MultiplayerQueueList : DrawableRoomPlaylist
+    {
+        public MultiplayerQueueList()
+            : base(false, false, true)
+        {
+        }
+
+        protected override FillFlowContainer<RearrangeableListItem<PlaylistItem>> CreateListFillFlowContainer() => new QueueFillFlowContainer
+        {
+            Spacing = new Vector2(0, 2)
+        };
+
+        private class QueueFillFlowContainer : FillFlowContainer<RearrangeableListItem<PlaylistItem>>
+        {
+            [Resolved(typeof(Room), nameof(Room.QueueMode))]
+            private Bindable<QueueMode> queueMode { get; set; }
+
+            [Resolved(typeof(Room), nameof(Room.Playlist))]
+            private BindableList<PlaylistItem> roomPlaylist { get; set; }
+
+            protected override void LoadComplete()
+            {
+                base.LoadComplete();
+                queueMode.BindValueChanged(_ => InvalidateLayout());
+            }
+
+            public override IEnumerable<Drawable> FlowingChildren
+            {
+                get
+                {
+                    switch (queueMode.Value)
+                    {
+                        default:
+                            return AliveInternalChildren.Where(d => d.IsPresent)
+                                                        .OfType<RearrangeableListItem<PlaylistItem>>()
+                                                        .OrderBy(item => item.Model.ID);
+
+                        case QueueMode.AllPlayersRoundRobin:
+                            RearrangeableListItem<PlaylistItem>[] items = AliveInternalChildren
+                                                                          .Where(d => d.IsPresent)
+                                                                          .OfType<RearrangeableListItem<PlaylistItem>>()
+                                                                          .OrderBy(item => item.Model.ID)
+                                                                          .ToArray();
+
+                            int totalCount = items.Length;
+                            if (totalCount == 0)
+                                return Enumerable.Empty<Drawable>();
+
+                            // Count of "expired" items per user.
+                            Dictionary<int, int> perUserCounts = roomPlaylist
```

Comment:
I'm confused why this is using "expired" counts, shouldn't it be the opposite? Otherwise in the case when there's a room with two users that are playing for hours, then joined by a third, would the first two users never get to choose again?

Also really nervous having this implemented here. Is the plan that this is temporary until the server sends proper ordering? Is there a reason that shouldn't be done as a priority?

Reply:
> Otherwise in the case when there's a room with two users that are playing for hours, then joined by a third, would the first two users never get to choose again?

Basically, yep. This is true for the server implementation too.

> temporary until the server sends proper ordering

Yeah, I suppose so. I'll look into that next.

Reply:
Definitely think the algorithm needs to change. That sounds like one of the reason people are confused / not using it currently.

Reply:
Going to turn a blind eye to this code section as it will be replaced in the near future.

Reply:
yeah, the current algorithm is a bit confusing. definitely needs some change and perhaps another queue mode that just alternates between everyone in the room would also be cool. but i think that's a topic for a separate PR or maybe discussion?

Reply:
Regardless that's not a discussion for this PR. I'll think about it server-side for a future PR.